src/mariadb_udf/config.c

```diff
@@ -0,0 +1,308 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+  
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
+  USA.
+*/
+#include "config.h"
+#include <string.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <errno.h>
+
+struct Lookup {
+    char* name;
+    int val;
+};
+
+// Lookup using binary search;
+int _lookup(char* name, ssize_t name_len, struct Lookup* lookup, int len)
+{
+    int low = 0;
+    while (len) {
+        int mid = low + len / 2;
+        int cmp = strncmp(name, lookup[mid].name, name_len);
+        if (cmp == 0) {
+            return lookup[mid].val;
+        } else if (cmp > 0) {
+            low = mid + 1;
+            len--;
+        }
+
+        len >>= 1;
+    }
+
+    return -1;
+}
+
+int setup_syslog_header(struct syslog_header* header, char* line, ssize_t len);
+
+int setup_cef_header(struct cef_header* header, char* line, ssize_t len);
+
+// Must be sorted by name
+static struct Lookup PRIORITY_LOOKUP[] = {
+    { "alert", 1 },
+    { "crit", 2 },
+    { "debug", 7 },
+    { "emerg", 0 },
+    { "err", 3 },
+    { "info", 6 },
+    { "notice", 5 },
+    { "warning", 4 },
+};
+
+// Must be sorted by name
+static struct Lookup FACILITY_LOOKUP[] = {
+    { "auth", 4 }, // Security/authentication messages
+    { "authpriv", 10 }, // Security/authentication messages
+    { "console", 14 }, // Log alert
+    { "cron", 9 }, // Clock daemon
+    { "daemon", 3 }, // System daemons
+    { "ftp", 11 }, // FTP daemon
+    { "kern", 0 }, // Kernel messages
+    { "local0", 16 }, // Locally used facility
+    { "local1", 17 }, // Locally used facility
+    { "local2", 18 }, // Locally used facility
+    { "local3", 19 }, // Locally used facility
+    { "local4", 20 }, // Locally used facility
+    { "local5", 21 }, // Locally used facility
+    { "local6", 22 }, // Locally used facility
+    { "local7", 23 }, // Locally used facility
+    { "lpr", 6 }, // Line printer subsystem
+    { "mail", 2 }, // Mail system
+    { "news", 7 }, // Network news subsystem
+    { "ntp", 12 }, // NTP subsystem
+    { "security", 13 }, // Log audit
+    { "solaris-cron", 15 }, // Scheduling daemon
+    { "syslog", 5 }, // Messages generated internally by syslogd
+    { "user", 1 }, // User-level messages
+    { "uucp", 8 }, // UUCP subsystem
+};
+
+int priority_lookup(char* name, ssize_t len)
+{
+    return _lookup(name, len, PRIORITY_LOOKUP, sizeof(PRIORITY_LOOKUP) / sizeof(PRIORITY_LOOKUP[0]));
+}
+
+int facility_lookup(char* name, ssize_t len)
+{
+    return _lookup(name, len, FACILITY_LOOKUP, sizeof(FACILITY_LOOKUP) / sizeof(FACILITY_LOOKUP[0]));
+}
+
+#define get_val(f, v, l)               \
+    do {                               \
+        char* tmp;                     \
+        char* field = strstr(line, f); \
```

Comment:
strstr is usually unsafe. strnstr should be used

Reply:
strnstr is not in the libc in Linux.
I used the equivalent function memmem.